File: sdk/go/arvados/group.go
// Copyright (C) The Arvados Authors. All rights reserved.
//
// SPDX-License-Identifier: Apache-2.0

package arvados

import (
	"time"
)

// Group is an arvados#group record
type Group struct {
	UUID                 string                 `json:"uuid"`
	Name                 string                 `json:"name"`
	OwnerUUID            string                 `json:"owner_uuid"`
	GroupClass           string                 `json:"group_class"`
	Etag                 string                 `json:"etag"`
	Href                 string                 `json:"href"`
	TrashAt              *time.Time             `json:"trash_at"`
	CreatedAt            time.Time              `json:"created_at"`
	ModifiedAt           time.Time              `json:"modified_at"`
	ModifiedByClientUUID string                 `json:"modified_by_client_uuid"`
	ModifiedByUserUUID   string                 `json:"modified_by_user_uuid"`
	DeleteAt             *time.Time             `json:"delete_at"`
	IsTrashed            bool                   `json:"is_trashed"`
	Properties           map[string]interface{} `json:"properties"`
	WritableBy           []string               `json:"writable_by,omitempty"`
	Description          string                 `json:"description"`
}

// GroupList is an arvados#groupList resource.
type GroupList struct {
	Items          []Group `json:"items"`
	ItemsAvailable int     `json:"items_available"`
	Offset         int     `json:"offset"`
	Limit          int     `json:"limit"`
}

// ObjectList is an arvados#objectList resource.
type ObjectList struct {
	Items          []interface{} `json:"items"`
	ItemsAvailable int           `json:"items_available"`
	Offset         int           `json:"offset"`
	Limit          int           `json:"limit"`
}

func (g Group) resourceName() string {
	return "group"
}

File: services/workbench2/src/models/group.ts
// Copyright (C) The Arvados Authors. All rights reserved.
//
// SPDX-License-Identifier: AGPL-3.0

import {
    Resource,
    ResourceKind,
    ResourceWithProperties,
    RESOURCE_UUID_REGEX,
    ResourceObjectType,
    TrashableResource
} from "./resource";

export interface GroupResource extends TrashableResource, ResourceWithProperties {
    kind: ResourceKind.GROUP;
    name: string;
    groupClass: GroupClass | null;
    description: string;
    ensure_unique_name: boolean;
    canWrite: boolean;
    canManage: boolean;
    // Optional local-only field, undefined for not loaded, null for failed to load
    memberCount?: number | null;
}

export enum GroupClass {
    PROJECT = 'project',
    FILTER = 'filter',
    ROLE = 'role',
}

export enum BuiltinGroups {
    ALL = 'fffffffffffffff',
    ANON = 'anonymouspublic',
    SYSTEM = '000000000000000',
}

export const getBuiltinGroupUuid = (cluster: string, groupName: BuiltinGroups): string => {
    return cluster ? `${cluster}-${ResourceObjectType.GROUP}-${groupName}` : "";
};

export const isBuiltinGroup = (uuid: string) => {
    const match = RESOURCE_UUID_REGEX.exec(uuid);
    const parts = match ? match[0].split('-') : [];
    return parts.length === 3 && parts[1] === ResourceObjectType.GROUP && Object.values<string>(BuiltinGroups).includes(parts[2]);
};

export const isUserGroup = (resource: any): resource is GroupResource => {
    return resource && resource.kind === ResourceKind.GROUP && resource.groupClass === GroupClass.ROLE;
};

export const isGroupResource = (resource: Resource): resource is GroupResource => {
    return resource && resource.kind === ResourceKind.GROUP;
};

export const selectedFieldsOfGroup = [
    "uuid",
    "name",
    "group_class",
    "description",
    "properties",
    "can_write",
    "can_manage",
    "trash_at",
    "delete_at",
    "is_trashed",
    "frozen_by_uuid"
];
